pipeline/04-interpret.R

```diff
@@ -105,6 +112,126 @@ lightgbm::lgb.importance(lgbm_final_full_fit$fit) %>%
+# 4. Calculate comps -----------------------------------------------------------
+#- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+if (comp_enable) {
+  message("Calculating comps")
+
+  # Calculate the leaf node assignments for every predicted value.
+  # Due to integer overflow problems with leaf node assignment, we need to
+  # chunk our data such that they are strictly less than the limit of 1073742
+  # rows. More detail here: https://github.com/microsoft/LightGBM/issues/1884
+  chunk_size <- 500000
+  chunks <- split(
+    assessment_data_prepped,
+    ceiling(seq_along(assessment_data_prepped[[1]]) / chunk_size)
+  )
+  chunked_leaf_nodes <- chunks %>%
+    map(\(chunk) {
+      predict(
+        object = lgbm_final_full_fit$fit,
+        newdata = as.matrix(chunk),
+        type = "leaf",
+      )
+    })
+  # Prefer do.call(rbind, ...) over bind_rows() because the chunks are
+  # not guaranteed to have the same number of rows, and bind_rows() will raise
+  # an error in that case
+  leaf_nodes <- do.call(rbind, chunked_leaf_nodes) %>% as_tibble()
+
+  # Calculate weights representing feature importance, so that we can weight
+  # leaf node assignments based on the most important features.
+  # To do this, we need the training data so that we can compute the mean sale
+  # price and use it as the base model error
+  message("Extracting weights from training data")
+  training_data <- read_parquet(paths$input$training$local) %>%
+    filter(!ind_pin_is_multicard, !sv_is_outlier) %>%
+    as_tibble()
+
+  tree_weights <- extract_weights(
+    model = lgbm_final_full_fit$fit,
+    mean_sale_price = mean(training_data[["meta_sale_price"]]),
+    metric = params$model$objective
+  )
+
+  # Get predicted values and leaf node assignments for the training data
+  training_data_prepped <- recipes::bake(
+    object = lgbm_final_full_recipe,
+    new_data = training_data,
+    all_predictors()
+  )
+  training_leaf_nodes <- predict(
+    object = lgbm_final_full_fit$fit,
+    newdata = as.matrix(training_data_prepped),
+    type = "leaf"
+  ) %>%
+    as_tibble()
+  training_leaf_nodes$predicted_value <- predict(
+    object = lgbm_final_full_fit$fit,
+    newdata = as.matrix(training_data_prepped)
+  ) %>%
+    # Round predicted values down for binning
+    floor()
+
+  # Get predicted values for the assessment set, which we already have in
+  # the assessment card set
+  leaf_nodes$predicted_value <- assessment_data %>%
+    left_join(assessment_card, by = c("meta_pin", "meta_card_num")) %>%
+    # Round predicted values down for binning
+    mutate(pred_card_initial_fmv = floor(pred_card_initial_fmv)) %>%
+    dplyr::pull("pred_card_initial_fmv")
+
+  # Make sure that the leaf node tibbles are all integers, which is what
+  # the comps algorithm expects
+  leaf_nodes <- leaf_nodes %>% mutate_all(as.integer)
+  training_leaf_nodes <- training_leaf_nodes %>% mutate_all(as.integer)
```

Comment:
This block ended up fixing the float-index slicing error that I was debugging yesterday. It seems like somehow the transformations we tweaked above ended up causing reticulate to interpret all numeric tibble values as floats rather than integers; I'm not 100% sure which transformation caused that change in behavior, but explicitly casting the tibbles to integers here resolves the issue.